packages/next/client/router.ts

```diff
@@ -38,7 +46,7 @@ const urlPropertyFields = [
   'isFallback',
   'basePath',
 ]
-const routerEvents = [
+const routerEvents: Array<RouterEvent> = [
```

Comment:
Isn't this automatically inferred by typescript 🤔

Reply:
Should be... But I don't why the auto-completion isn't working. 

```
routerEvents.forEach((event) => {
  singletonRouter.ready(() => {
    Router.events.on(event, (...args) => {
```

Reply:
Had a look into this, the problem was that we use mitt without a generic type, so it would always be a set "string" type instead of the actual values.

I've pushed up a commit that adds a generic for mitt and then provides the values to it: [`0476b70` (#15497)](https://github.com/vercel/next.js/pull/15497/commits/0476b7028f04ac7190d49b282569724ff85123ab)

This only solves the first parameter tbh, you could take it a bit further by adding a generic for the function itself as well. E.g. the type autocomplete could be `'routeChangeComplete', (url: string) => void`

Reply:
Thanks @timneutkens, really appreciate your help 😄 

> This only solves the first parameter tbh, you could take it a bit further by adding a generic for the function itself as well. E.g. the type autocomplete could be 'routeChangeComplete', (url: string) => void

Sure! I'll definitely send a commit for it 

---

Btw, instead of repeating the same strings as in the ``routerEvents`` arrays just to declare a ``RouterEvent`` literal type, shouldn't be better to create a helper to handle this?

For instance:
```
const asLiterals = <T extends string>(array: T[]): T[] => array
```

```
const routerEvents: asLiterals([...])
```

Reply:
Overall that would affect bundle size in a negative way as it introduces an extra function

Reply:
@timneutkens Just added a generic for the event handler in 3483e1ab68f074230b862839582b9561cab8d546 and 1d76bdf169d9ba3e089fb46053d1fab55be12e81 

What do you think?

Reply:
@timneutkens Just added the router events map, what do you think?